NO_TICKET | Remodeling DTO outputs, delegating to specific services for every use case category

VictorDelCampo opened the pull request:
# Short Description
- Starting point --> Output DTOs were under the API layer
- Ending point --> Output DTOs are under the application layer and the response creation is done by specific services. This way the API layer doesn't know about the DTO output creation

# Changes
- Moved Output DTOs to the application layer, together with their corresponding input DTOs
- Created new services for every use case category, so that the output DTO creation is done in the application layer and not longer in the API layer.
- test_api updated so that the correct mock is done for the repository.

# Feedback
- Wdyt?

# How to review this Pull Request
[Link to Confluence](https://digitalservice4germany.atlassian.net/wiki/spaces/STL/pages/117637157/Development+Process#Approve-or-comment) (internal)

VictorDelCampo commented:
> It's nice that you now want to introduce individual services with interface segregation again to step by step getting back to the code/architecture that we already had at the beginning! Thx!

Yes! This way the API layer wouldn't have to create the response DTOS no more!

punknoir101 commented:
> Yes! This way the API layer wouldn't have to create the response DTOS no more!

Exactly.

VictorDelCampo commented:
@JulianRoesner @SannyNguyenHung could you also have a look?

SannyNguyenHung commented on erica/api/v2/endpoints/est.py:
```diff
@@ -40,8 +45,8 @@ async def get_send_est_job(request_id: UUID):
     try:
-        erica_request = get_erica_request(request_id)
-        return create_response(erica_request)
+        tax_declaration_service: TaxDeclarationServiceInterface = injector.inject(TaxDeclarationServiceInterface)
```
Why do we moved the injector into the class? The injector the send_est should not use the injector directly instead we should use a method or a parameter. Wdyt?

VictorDelCampo replied:
We had the injector before in the utils class and we were calling from the API endpoints that utils class. But now we have a service interface for every use case and therefore it is (for now) being injected in every endpoint.
But we could maybe moved that service definition to the injector factory?

JulianRoesner commented on erica/api/v2/endpoints/est.py:
```diff
@@ -2,21 +2,26 @@
 
 router = APIRouter()
 
+injector = Injector([
+    ApiModule(),
+])
```
I am not the biggest fan of that. Because if we hardcode it, we do not really have a possibility to change it for e.g. the tests and then have to go back to mocking. I am wondering if there is a way to get that from somewhere else (s. also the comment from Sanny)

JulianRoesner commented on erica/api/v2/responses/model.py:
```diff
@@ -1,99 +1,48 @@
     200: {"model": EstResponseDto,
           "description": "Job status of a sent est was successfully retrieved from the queue."},
-    500: model_error_request_queue}
+    500: model_error_get_from_queue}
```
Isn't the 404 missing?

JulianRoesner commented on erica/application/tax_declaration/TaxDeclarationService.py:
```diff
@@ -0,0 +1,46 @@
+from erica.application.tax_declaration.tax_declaration_dto import EstResponseDto, ResultEstResponseDto
+
+
+class TaxDeclarationServiceInterface:
+    __metaclass__ = ABCMeta
+
+    @abstractmethod
+    def get_response_send_est(self, request_id: UUID) -> EstResponseDto:
+        pass
```
If all those Interfaces had one common base class, we could put  `erica_request_service.get_request_by_request_id(request_id)` and error handling for the JobTypeDoesNotMatchEndpointError into that base class and use it for all of the different services. Wdyt?

VictorDelCampo replied:
I have just added a base_service, have a look!

JulianRoesner reviewed (commented):
Nothing that would prevent that from being merged. Just some comments where I would see potential improvements

VictorDelCampo commented on erica/api/v2/responses/model.py:
```diff
@@ -1,99 +1,48 @@
 
 response_model_get_send_est_from_queue = {
     200: {"model": EstResponseDto,
           "description": "Job status of a sent est was successfully retrieved from the queue."},
-    404: model_error_request_queue,
-    500: model_error_request_queue}
+    404: model_error_get_from_queue,
+    500: model_error_get_from_queue}
```
So much cleaner!

VictorDelCampo commented on erica/application/FreischaltCode/FreischaltCodeService.py:
```diff
@@ -0,0 +1,71 @@
+        super().__init__()
+        self.erica_request_service = service
+
+    def get_response_freischaltcode_request_and_activation(self, request_id: UUID):
+        erica_request = self.erica_request_service.get_request_by_request_id(request_id)
+        process_status = map_status(erica_request.status)
+        if process_status == JobState.SUCCESS:
+            result = ResultFreischaltcodeRequestAndActivation(
+                elster_request_id=erica_request.result["elster_request_id"],
+                transfer_ticket=erica_request.result["transfer_ticket"],
+                idnr=erica_request.payload.get("idnr"))
+            return FreischaltcodeRequestAndActivationResponseDto(
+                processStatus=map_status(erica_request.status), result=result)
+        elif process_status == JobState.FAILURE:
+            return FreischaltcodeRequestAndActivationResponseDto(
+                processStatus=map_status(erica_request.status), errorCode=erica_request.error_code,
+                errorMessage=erica_request.error_message)
+        else:
+            return FreischaltcodeRequestAndActivationResponseDto(
+                processStatus=map_status(erica_request.status))
```
Looks good :)

VictorDelCampo commented on erica/application/tax_number_validation/TaxNumberValidityService.py:
```diff
@@ -0,0 +1,45 @@
+                errorMessage=erica_request.error_message)
+        else:
+            return TaxResponseDto(
+                processStatus=map_status(erica_request.status))
```
Tests added

JulianRoesner commented on tests/application/tax_declaration/test_tax_declaration_service.py:
```diff
@@ -0,0 +1,61 @@
+
+class TestTaxDeclarationService:
+
+    @pytest.mark.parametrize("status", [Status.new, Status.scheduled, Status.processing])
```
Love that use of the parametrize 👍 

And thanks for those tests 😍

JulianRoesner commented on erica/application/Shared/base_service.py:
```diff
@@ -0,0 +1,13 @@
+
+class BaseService:
+    erica_request_service: EricaRequestService
+
```
Could we add a type here, that is None for the base class and would be RequestType.freischaltcode_activate etc. for the other classes. But we can also add that in a separate PR

VictorDelCampo replied:
What do we need type for in the subclasses?

JulianRoesner replied:
This was just an idea how we couldset the correct request types for all the different request. However, we could also put this as a parameter into the function. Nevertheless, this probably does not make sense to do here but directly when we implement raising an error in case the type was not found.

VictorDelCampo replied:
oh true you we thinking a bit further regarding your PR with the error handling. Let me take over in your PR after we merge this one

JulianRoesner reviewed (commented):
Nice 👍

VictorDelCampo commented:
@SannyNguyenHung maybe you could have another quick look? I have removed the injectors from the API lever and introduced a service injector within the application layer. I also refactored the test_api so that they do not depend on a running database no more

SannyNguyenHung commented on erica/application/Shared/service_injector.py:
```diff
@@ -0,0 +1,24 @@
+        RequestType.freischalt_code_activate: injector.inject(FreischaltCodeServiceInterface),
+        RequestType.freischalt_code_revocate: injector.inject(FreischaltCodeServiceInterface),
+        RequestType.check_tax_number: injector.inject(TaxNumberValidityServiceInterface),
+        RequestType.send_est: injector.inject(TaxDeclarationServiceInterface),
```
can you add grundsteuer here?

VictorDelCampo replied:
Done!

SannyNguyenHung reviewed (approved):
LGTM! But don't forget the grundsteuer, maybe you should merge main into your branch

VictorDelCampo commented:
@SannyNguyenHung I have just integrated the main state into this feature branch and upgraded the grundsteuer API. Tests are also updated. Could you have another quick look?
